### PR TITLE
crypto: polyval - Add POLYVAL support

### PR DESCRIPTION
**Pull request by nhukc**
- Add POLYVAL support

### CODE REVIEW

File: crypto/algapi.c [hunk 1]
```diff
@@ -1028,6 +1028,61 @@ void __crypto_xor(u8 *dst, const u8 *src1, const u8 *src2, unsigned int len)
 }
 EXPORT_SYMBOL_GPL(__crypto_xor);
 
+// Computes dst = src1 ^ strrev(src2)
+void __crypto_xor_reverse(u8 *dst, const u8 *src1, const u8 *src2, unsigned int len)
```
Comment (nhukc) [inline crypto/algapi.c]: Modify this code to only perform the byteswap and move it to `polyval.c`. Calling the byteswap, then `crypto_xor` would have the same effect and would be cleaner.

File: crypto/polyval-generic.c [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+static int polyval_init(struct shash_desc *desc)
+{
+	struct polyval_desc_ctx *dctx = shash_desc_ctx(desc);
+
+	memset(dctx, 0, sizeof(*dctx));
+
+	return 0;
+}
+
+static void reverse_bytes(u8 *buf) {
```
Comment (nhukc) [inline crypto/polyval-generic.c]: This code and `crypto_xor_reverse` could be simplified into a single function that efficiently reverses an arbitrary length bytestring.

File: crypto/testmgr.c [hunk 3]
```diff
@@ -5257,6 +5257,13 @@ static const struct alg_test_desc alg_test_descs[] = {
 		.suite = {
 			.hash = __VECS(poly1305_tv_template)
 		}
+	}, {
+		.alg = "polyval",
+		.test = alg_test_hash,
+		.fips_allowed = 1,
```
Comment (ebiggers) [inline crypto/testmgr.c]: `fips_allowed` shouldn't be set.

File: crypto/testmgr.h [hunk 4]
```diff
@@ -32800,4 +32800,206 @@ static const struct hash_testvec blakes2s_256_tv_template[] = {{
 			  0xd5, 0x06, 0xb5, 0x3a, 0x7c, 0x7a, 0x65, 0x1d, },
 }};
 
+static const struct hash_testvec polyval_tv_template[] = {
```
Comment (ebiggers) [inline crypto/testmgr.h:32803]: There should be a comment here that says where these test vectors came from.

File: crypto/testmgr.h [hunk 5]
```diff
@@ -32800,4 +32800,206 @@ static const struct hash_testvec blakes2s_256_tv_template[] = {{
...
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x04\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x05\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x08\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x02\x00\x00\x00\x00\x00\x00",
+		.digest	= "\xff\xcd\x05\xd5\x77\x0f\x34\xad"
+			  "\x92\x67\xf0\xa5\x99\x94\xb1\x5a",
+		.psize	= 96,
```
Comment (ebiggers) [inline crypto/testmgr.h:32951]: There should be a test vector that's a bit longer, like 256 bytes.  Some implementations might not get properly tested without a longer input.  It would be okay to delete a couple of the current test vectors to "make room" for it, especially ones that use duplicate lengths, as those aren't too useful.

File: include/crypto/algapi.h [hunk 6]
```diff
@@ -180,6 +181,23 @@ static inline void crypto_xor_cpy(u8 *dst, const u8 *src1, const u8 *src2,
 	}
 }
 
+static inline void crypto_xor_reverse(u8 *dst, const u8 *src, unsigned int size)
```
Comment (ebiggers) [inline include/crypto/algapi.h]: Why is `crypto_xor_reverse()` needed?  If this is needed, it should be added in a separate patch.

If this is just an optimization to avoid copying each block of data to a stack buffer, I don't think you should worry about that too much.  This is just the generic implementation, and people will use one the architecture-optimized implementation if they care about performance.

File: include/crypto/polyval.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+#define POLYVAL_BLOCK_SIZE	16
+#define POLYVAL_DIGEST_SIZE	16
+
+struct polyval_ctx {
+	struct gf128mul_4k *gf128;
+};
+
+struct polyval_desc_ctx {
+	u8 buffer[POLYVAL_BLOCK_SIZE];
```
Comment (ebiggers) [inline include/crypto/polyval.h:16]: Weird indentation here.  Make sure you run `./scripts/checkpatch.pl`; it will complain about things like this.

File: crypto/polyval-generic.c [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	.update		= polyval_update,
+	.final		= polyval_final,
+	.setkey		= polyval_setkey,
+	.descsize	= sizeof(struct polyval_desc_ctx),
+	.base		= {
+		.cra_name		= "polyval",
+		.cra_driver_name	= "polyval-generic",
+		.cra_priority		= 100,
+		.cra_blocksize		= POLYVAL_BLOCK_SIZE,
+		.cra_ctxsize		= sizeof(struct polyval_ctx),
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: I know that `ghash-generic.c` does this with `ghash_ctx`, but the naming of `polyval_ctx` is a bit confusing.  It really should be `polyval_tfm_ctx` or `polyval_key`.

File: crypto/polyval-generic.c [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    reverse_bytes(buf);
+	memcpy(dst, buf, POLYVAL_BLOCK_SIZE);
+
+	return 0;
+}
+
+static void polyval_exit_tfm(struct crypto_tfm *tfm)
+{
+	struct polyval_ctx *ctx = crypto_tfm_ctx(tfm);
+	if (ctx->gf128)
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: This NULL check is unnecessary.

File: crypto/polyval-generic.c (hunk 2 above)
Comment (ebiggers) [reply crypto/polyval-generic.c]: Why does it need to be arbitary length?  It looks like in both cases, only `POLYVAL_BLOCK_SIZE` bytes are reversed.

I think something like the following that's specialized for 16-byte blocks would be better suited for this:

```
static void reverse_block(u8 block[POLYVAL_BLOCK_SIZE])                          
{                                                                                
        u64 *p1 = (u64 *)block;                                                  
        u64 *p2 = (u64 *)&block[8];                                              
        u64 a = get_unaligned(p1);                                               
        u64 b = get_unaligned(p2);                                               
                                                                                 
        put_unaligned(swab64(a), p2);                                            
        put_unaligned(swab64(b), p1);                                            
}                                                                                
```

File: crypto/polyval-generic.c [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+static int polyval_setkey(struct crypto_shash *tfm,
+			const u8 *key, unsigned int keylen)
+{
+	struct polyval_ctx *ctx = crypto_shash_ctx(tfm);
+	be128 k;
+
+	if (keylen != POLYVAL_BLOCK_SIZE)
+		return -EINVAL;
+
+	if (ctx->gf128)
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: This NULL check is unnecessary.

File: crypto/Kconfig [hunk 11]
```diff
@@ -751,6 +751,14 @@ config CRYPTO_GHASH
 	  GHASH is the hash function used in GCM (Galois/Counter Mode).
 	  It is not a general-purpose cryptographic hash function.
 
+config CRYPTO_POLYVAL
+	tristate "POLYVAL hash function"
```
Comment (ebiggers) [inline crypto/Kconfig]: I don't think there's much reason to make this option user-selectable, as opposed to just making it be selected by `CRYPTO_HCTR2`, and by the architecture-specific POLYVAL implementations if they'll rely on this as a fallback.  (Having a prompt string after `tristate` makes it user-selectable; having just `tristate` makes it non-user-selectable.)  See what's done for `CRYPTO_NHPOLY1305`, for example.

File: crypto/polyval-generic.c [hunk 12]
```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * POLYVAL: hash function for GCM-SIV.
+ * Code based on crypto/ghash-generic.c
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: Can you add a comment that briefly explains how this implementation works?  Especially how it differs from ghash-generic.

File: crypto/polyval-generic.c [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+		dctx->bytes = POLYVAL_BLOCK_SIZE - srclen;
+		pos = dst + POLYVAL_BLOCK_SIZE - 1;
+		while (srclen--)
+			*pos-- ^= *src++;
+	}
+
+	return 0;
+}
+
+static void polyval_flush(struct polyval_ctx *ctx, struct polyval_desc_ctx *dctx)
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: Is it worth having this as a separate function, instead of just be part of `polyval_final()`?

File: crypto/polyval-generic.c [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+		return -ENOMEM;
+
+	return 0;
+}
+
+static int polyval_update(struct shash_desc *desc,
+			 const u8 *src, unsigned int srclen)
+{
+	struct polyval_desc_ctx *dctx = shash_desc_ctx(desc);
+	struct polyval_ctx *ctx = crypto_shash_ctx(desc->tfm);
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: `polyval_ctx` should be declared as `const` whenever possible, since it shouldn't be modified outside of `setkey`.

File: crypto/polyval-generic.c [hunk 15]
```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * POLYVAL: hash function for GCM-SIV.
+ * Code based on crypto/ghash-generic.c
+ *
+ * Copyright (c) 2007 Nokia Siemens Networks - Mikko Herranen <mh1@iki.fi>
+ * Copyright (c) 2009 Intel Corp.
+ *   Author: Huang Ying <ying.huang@intel.com>
+ * Copyright (c) 2021 Google LLC
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: No `(c)`, just use `Copyright 2021 Google LLC`

**Comment by nhukc** [comment]
Needs to wait for approval on https://github.com/google/hctr2/pull/42

File: crypto/ghash-generic.c [hunk 16]
```diff
@@ -61,8 +61,7 @@ static int ghash_setkey(struct crypto_shash *tfm,
 	if (keylen != GHASH_BLOCK_SIZE)
 		return -EINVAL;
 
-	if (ctx->gf128)
-		gf128mul_free_4k(ctx->gf128);
+	gf128mul_free_4k(ctx->gf128);
```
Comment (ebiggers) [inline crypto/ghash-generic.c]: This cleanup in ghash-generic.c should be a separate patch.  How about sending it out now as its own patch?  It doesn't need to wait for anything else.

File: crypto/polyval-generic.c [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+ * Code based on crypto/ghash-generic.c
+ *
+ * POLYVAL is a keyed hash function similar to GHASH. POLYVAL uses a
+ * different modulus for finite field multiplication which makes hardware
+ * accelerated implementations on little-endian machines faster.
+ *
+ * Like GHASH, POLYVAL is not a cryptographic hash function and should
+ * not be used outside of crypto modes explicitly designed to use POLYVAL.
+ *
+ * Copyright (c) 2007 Nokia Siemens Networks - Mikko Herranen <mh1@iki.fi>
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: Copyright lines should be closer to the top of the file.  See crypto/adiantum.c for a good way to do it.  Order is SPDX license identifier, then a block comment containing a 1-line summary of the file and the copyright line(s), and then a separate block comment with a long description of the file.

File: crypto/polyval-generic.c (hunk 10 above)
Comment (ebiggers) [reply crypto/polyval-generic.c]: This wasn't resolved.

File: crypto/polyval-generic.c [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+static int polyval_update(struct shash_desc *desc,
+			 const u8 *src, unsigned int srclen)
+{
+	struct polyval_desc_ctx *dctx = shash_desc_ctx(desc);
+	const struct polyval_tfm_ctx *ctx = crypto_shash_ctx(desc->tfm);
+	u8 *dst = dctx->buffer;
+	u8 *pos;
+    u8 tmp[POLYVAL_BLOCK_SIZE];
```
Comment (ebiggers) [inline crypto/polyval-generic.c]: There is still weird indentation here.  Make sure to run `checkpatch`.

**Review by ebiggers — COMMENTED** [review]
Can you squash the second commit into the first one?
